File: klinara-ios/klinara-ios/Services/Packages/PackageOperationModels.swift
import Foundation

// Kaynak: `apps/api/src/modules/packages/dto/package-operation.dto.ts`.
//
// Bu dosyadaki gövdelerin üçü (`adjust`, `refund`, `transfer`) hem `If-Match`
// hem `Idempotency-Key` ister ve bu KASITLI: ilki "bayat durum üzerinde işlem
// yaptın"ı, ikincisi "aynı isteği tekrar gönderdin"i durdurur. Biri diğerinin
// yerini tutmaz.

// MARK: - Kullanılabilir haklar

/// `PackageEntitlementDto` — randevu ekranının paket seçimi için.
///
/// Sunucu yalnız **aktif, süresi dolmamış ve kalanı olan** kalemleri döner;
/// istemcinin ayrıca elemesi gerekmez.
nonisolated struct PackageEntitlement: Decodable, Sendable, Identifiable, Equatable {
    let customerPackageItemId: String
    let customerPackageId: String
    let packageName: String
    let serviceId: String
    let serviceName: String
    let remainingSessions: Int
    let expiresAt: Date?
    let branchId: String

    /// Liste kimliği kalem kimliğidir: bir müşteride aynı hizmetin iki ayrı
    /// paketten hakkı olabilir, paket kimliği tekil olmazdı.
    var id: String { customerPackageItemId }
}

// MARK: - Düzeltme

nonisolated struct AdjustItemInput: Encodable, Sendable, Equatable {
    let customerPackageItemId: String
    /// Pozitif = hak ekle, negatif = hak düş. **Sıfır olamaz.**
    let delta: Int
}

/// `AdjustPackageDto` — gerekçe ZORUNLU, sunucu ve veritabanı da zorluyor.
nonisolated struct AdjustPackageInput: Encodable, Sendable, Equatable {
    let items: [AdjustItemInput]
    let reason: String

    /// Sunucudaki `@MinLength(5)` kuralının aynası — kullanıcıya `422` yerine
    /// pasif bir Kaydet düğmesi göstermek için.
    static let minimumReasonLength = 5

    var isValid: Bool {
        !items.isEmpty
            && items.allSatisfy { $0.delta != 0 }
            && reason.trimmingCharacters(in: .whitespacesAndNewlines).count
                >= Self.minimumReasonLength
    }
}

// MARK: - İade

nonisolated struct RefundItemInput: Encodable, Sendable, Equatable {
    let customerPackageItemId: String
    /// İade edilecek seans sayısı, en az 1.
    let sessions: Int
}

/// `RefundPackageDto` — `items` verilmezse TÜM kalan hak iade edilir.
nonisolated struct RefundPackageInput: Encodable, Sendable, Equatable {
    var items: [RefundItemInput]?
    let reason: String

    var isValid: Bool {
        let hasValidItems = items.map { !$0.isEmpty && $0.allSatisfy { $0.sessions > 0 } } ?? true
        return hasValidItems
            && reason.trimmingCharacters(in: .whitespacesAndNewlines).count
                >= AdjustPackageInput.minimumReasonLength
    }
}

/// `RefundResultDto`.
///
/// **Kasa hareketi yoktur.** Tutar satış anındaki tahsisten hesaplanır ve
/// yükümlülük `pending` yazılır; tahsilat tarafı Batch 6.2'de bağlanacak.
nonisolated struct RefundResult: Decodable, Sendable, Equatable {
    let refundedSessions: Int
    let refundAmountMinor: Int
    /// `pending` ya da `settled`.
    let settlementStatus: String
}

// MARK: - Devir

nonisolated struct TransferItemInput: Encodable, Sendable, Equatable {
    let customerPackageItemId: String
    let sessions: Int
}

/// `TransferPackageDto` — `items` verilmezse tüm kalan hak devredilir.
nonisolated struct TransferPackageInput: Encodable, Sendable, Equatable {
    let targetCustomerId: String
    var items: [TransferItemInput]?
    let reason: String

    var isValid: Bool {
        let hasValidItems = items.map { !$0.isEmpty && $0.allSatisfy { $0.sessions > 0 } } ?? true
        return !targetCustomerId.isEmpty
            && hasValidItems
            && reason.trimmingCharacters(in: .whitespacesAndNewlines).count
                >= AdjustPackageInput.minimumReasonLength
    }
}

// MARK: - Tüketim

nonisolated struct ConsumePackageLineInput: Encodable, Sendable, Equatable {
    /// Randevunun hizmet kalemi (`AppointmentService.id`).
    let appointmentServiceId: String
    /// Düşülecek müşteri paketi kalemi.
    let customerPackageItemId: String
}

/// `ConsumePackageDto` — randevu kalemlerini pakete bağlar.
nonisolated struct ConsumePackageInput: Encodable, Sendable, Equatable {
    let lines: [ConsumePackageLineInput]
}

/// `ConsumePackageResultDto`.
///
/// `consumed` randevu **henüz `completed` değilse 0**'dır ve bu bir hata
/// değildir: bağlama yapıldı, düşme randevu tamamlandığında aynı transaction
/// içinde kendiliğinden olacak.
nonisolated struct ConsumePackageResult: Decodable, Sendable, Equatable {
    let bound: Int
    let consumed: Int
}

File: klinara-android/app/src/main/java/com/klinara/android/services/packages/PackageOperationModels.kt
package com.klinara.android.services.packages

import com.klinara.android.services.networking.InstantSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.addJsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import java.time.Instant

// Kaynak: `apps/api/src/modules/packages/dto/package-operation.dto.ts`.

/**
 * `PackageEntitlementDto` — randevu ekranının paket seçimi için kullanılabilir hak.
 *
 * Sunucu yalnız **aktif, süresi dolmamış ve kalanı olan** kalemleri döner; istemcinin
 * ayrıca elemesi gerekmez. Liste kimliği KALEM kimliğidir: bir müşteride aynı hizmetin iki
 * ayrı paketten hakkı olabilir.
 */
@Serializable
data class PackageEntitlement(
    val customerPackageItemId: String,
    val customerPackageId: String,
    val packageName: String = "",
    val serviceId: String,
    val serviceName: String = "",
    val remainingSessions: Int = 0,
    @Serializable(with = InstantSerializer::class)
    val expiresAt: Instant? = null,
    val branchId: String,
)

/** Randevunun bir hizmet kalemini bir paket kalemine bağlar. */
data class ConsumePackageLineInput(
    /** Randevunun hizmet satırı (`AppointmentServiceLine.id`). */
    val appointmentServiceId: String,
    val customerPackageItemId: String,
)

/** `ConsumePackageDto`. */
data class ConsumePackageInput(
    val lines: List<ConsumePackageLineInput>,
) {
    fun toJson(): JsonObject =
        buildJsonObject {
            putJsonArray("lines") {
                lines.forEach { line ->
                    addJsonObject {
                        put("appointmentServiceId", line.appointmentServiceId)
                        put("customerPackageItemId", line.customerPackageItemId)
                    }
                }
            }
        }
}

/**
 * `ConsumePackageResultDto`.
 *
 * [consumed] randevu **henüz `completed` değilse 0'dır ve bu bir hata DEĞİL**: bağlama
 * yapıldı, düşme randevu tamamlandığında aynı transaction içinde kendiliğinden olacak.
 */
@Serializable
data class ConsumePackageResult(
    val bound: Int = 0,
    val consumed: Int = 0,
)

// --- A5.3: düzeltme, iade, devir ---
//
// Sunucu sözleşmesi (`package-operations.controller.ts`): üçü de `If-Match` İSTER; `refund`
// ve `transfer` ayrıca `Idempotency-Key` alır, `adjust` ALMAZ. İlki "bayat durum üzerinde
// işlem yaptın"ı, ikincisi "aynı isteği tekrar gönderdin"i durdurur — biri diğerinin
// yerini tutmaz. (iOS modelindeki "üçü de ikisini ister" notu `adjust` için yanlıştı.)

/** Gerekçe kuralı — sunucudaki `@MinLength(5) @MaxLength(500)` aynası. */
object OperationReason {
    const val MIN_LENGTH = 5
    const val MAX_LENGTH = 500

    fun isValid(reason: String): Boolean = reason.trim().length in MIN_LENGTH..MAX_LENGTH
}

/** Pozitif = hak ekle, negatif = hak düş. **Sıfır olamaz.** */
data class AdjustItemInput(
    val customerPackageItemId: String,
    val delta: Int,
)

/** `AdjustPackageDto` — gerekçe ZORUNLU; sunucu ve veritabanı da zorluyor. */
data class AdjustPackageInput(
    val items: List<AdjustItemInput>,
    val reason: String,
) {
    /** Kullanıcıya `400` yerine pasif bir "Uygula" düğmesi göstermek için. */
    val isValid: Boolean
        get() = items.isNotEmpty() && items.all { it.delta != 0 } && OperationReason.isValid(reason)

    fun toJson(): JsonObject =
        buildJsonObject {
            putJsonArray("items") {
                items.forEach { item ->
                    addJsonObject {
                        put("customerPackageItemId", item.customerPackageItemId)
                        put("delta", item.delta)
                    }
                }
            }
            put("reason", reason.trim())
        }
}

/** İade ya da devir için kalem başına seans sayısı (en az 1). */
data class SessionsItemInput(
    val customerPackageItemId: String,
    val sessions: Int,
)

/**
 * `RefundPackageDto`. [items] `null` ise TÜM kalan hak iade edilir — tam iadede liste
 * gönderilmez, sunucu kendisi hesaplar ve arada yarış olmaz.
 */
data class RefundPackageInput(
    val items: List<SessionsItemInput>?,
    val reason: String,
) {
    val isValid: Boolean
        get() =
            (items == null || (items.isNotEmpty() && items.all { it.sessions > 0 })) &&
                OperationReason.isValid(reason)

    fun toJson(): JsonObject =
        buildJsonObject {
            items?.let { putSessions(it) }
            put("reason", reason.trim())
        }
}

/**
 * `RefundResultDto`.
 *
 * **Kasa hareketi yoktur.** Tutar satış anındaki tahsisten hesaplanır ve yükümlülük
 * `pending` yazılır; tahsilat tarafı Faz A6'da bağlanacak.
 */
@Serializable
data class RefundResult(
    val refundedSessions: Int = 0,
    val refundAmountMinor: Long = 0,
    /** `pending` ya da `settled`. */
    val settlementStatus: String = "pending",
)

/** `TransferPackageDto` — [items] `null` ise tüm kalan hak devredilir. */
data class TransferPackageInput(
    val targetCustomerId: String,
    val items: List<SessionsItemInput>?,
    val reason: String,
) {
    val isValid: Boolean
        get() =
            targetCustomerId.isNotEmpty() &&
                (items == null || (items.isNotEmpty() && items.all { it.sessions > 0 })) &&
                OperationReason.isValid(reason)

    fun toJson(): JsonObject =
        buildJsonObject {
            put("targetCustomerId", targetCustomerId)
            items?.let { putSessions(it) }
            put("reason", reason.trim())
        }
}

private fun kotlinx.serialization.json.JsonObjectBuilder.putSessions(items: List<SessionsItemInput>) {
    putJsonArray("items") {
        items.forEach { item ->
            addJsonObject {
                put("customerPackageItemId", item.customerPackageItemId)
                put("sessions", item.sessions)
            }
        }
    }
}
